packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp

```diff
@@ -455,6 +460,72 @@ MakeCompositeDirectSolver(RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal
   return coarseSolver;
 } // MakeCorseCompositeDirectSolver
 
+/* Rebalance coarse operator
+ *
+ */
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void RebalanceCoarseCompositeOperator(const int maxRegPerProc,
+              const int rebalanceNumPartitions,
+              RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >& coarseCompOp,
+              RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> >& compCoarseCoordinates,
+              RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >& rebalancedCompOp,
+              RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> >& rebalancedCoordinates,
+              RCP<const Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> >& rebalanceImporter)
+{
+#include "MueLu_UseShortNames.hpp"
+  using CoordType = typename Teuchos::ScalarTraits<Scalar>::coordinateType;
+
+  RCP<Teuchos::FancyOStream> fos = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+  fos->setOutputToRootOnly(0);
+  *fos << "Rebalancing coarse composite operator." << std::endl;
+
+  RCP<const Map> map = coarseCompOp->getMap();
+
+  RCP<Xpetra::Vector<GlobalOrdinal,LocalOrdinal,GlobalOrdinal,Node> > decomposition = Xpetra::VectorFactory<GlobalOrdinal,LocalOrdinal,GlobalOrdinal,Node>::Build(map, true);
+
+  const int numPartitions = rebalanceNumPartitions;
+
+  // We build a fake level
+  Level level;
+  level.SetLevelID(1);
+
+  RCP<FactoryManagerBase> factoryHandler = rcp(new FactoryManager());
+  level.SetFactoryManager(factoryHandler);
+
+  RCP<ZoltanInterface> zoltan = rcp(new ZoltanInterface());
```

Comment:
Does this work with Tpetra?

Reply:
It does, unless I have tricked myself into not actually using Tpetra. But should change this to zoltan2 right?

Reply:
I be believe that Zoltan might have predated Epetra and it could rely only on plain old data pointers... but Zoltan2 would make more sense.